File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
// DFS / RECURSIVE SOLUTION
// const maxDepth = (root) => {
//     if (!root) return 0;
//     return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
// };

// BFS / ITERATIVE SOLUTION
const maxDepth = (root) => {
  if (!root) return 0;
  const queue = [root];
  let depth = 0;
  while (queue.length) {
    const len = queue.length;
    for (let i = 0; i < len; i++) {
      const node = queue.shift();
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    depth++;
  }
  return depth;
}

File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0;
        stack = [root]
        level = 0
        while stack:
            tempStack = []
            level += 1
            while stack:
                node = stack.pop()
                if node.left:
                    tempStack.append(node.left)
                if node.right:
                    tempStack.append(node.right)
            stack = tempStack
        return level